### PR TITLE
UCP: handle error on zcopy message

### PR DESCRIPTION
**Pull request by evgeny-leksikov**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3749/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1808/ for details.

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1815/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3755/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1827/ for details.

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3766/ for details (Mellanox internal link).

File: src/ucp/tag/tag_send.c [hunk 1]
```diff
@@ -218,6 +218,7 @@ static void ucp_tag_send_req_init(ucp_request_t* req, ucp_ep_h ep,
     req->send.buffer       = buffer;
     req->send.datatype     = datatype;
     req->send.tag          = tag;
+    req->send.rsc          = UCP_NULL_RESOURCE;
```
Comment (yosefe) [inline src/ucp/tag/tag_send.c]: let's move it under ENABLE_ASSERT

File: src/ucp/tag/rndv.c [hunk 2]
```diff
@@ -29,8 +29,9 @@ static int ucp_tag_rndv_is_get_op_possible(ucp_ep_h ep, uct_rkey_t rkey)
 
 static void ucp_rndv_rma_request_send_buffer_dereg(ucp_request_t *sreq)
 {
-    if ((UCP_DT_IS_CONTIG(sreq->send.datatype)) &&
-        (ucp_ep_is_rndv_lane_present(sreq->send.ep))) {
+    if (UCP_DT_IS_CONTIG(sreq->send.datatype) &&
+        ucp_ep_is_rndv_lane_present(sreq->send.ep) &&
+        ucp_request_is_send_buffer_reg(sreq) /* TODO: How is this possible? */) {
```
Comment (yosefe) [inline src/ucp/tag/rndv.c]: why did you add this?

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: There is a deregistration of non registered memory or registered but from very non-obvious place. It looks like protocol issue for me.

File: src/ucp/tag/tag_send.c (hunk 1 above)
Comment (evgeny-leksikov) [inline src/ucp/tag/tag_send.c]: See your comment below, it may doesn't work

File: src/ucp/tag/rndv.c (hunk 2 above)
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: it's happening on MM transport

Comment (yosefe) [inline src/ucp/tag/rndv.c]: is it a result of error handling? because it used to work before
adding @alinask

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: It's result of caching of `rsc` in request. before the info was taken from config on deregistration

Comment (yosefe) [inline src/ucp/tag/rndv.c]: but caching rsc should not affect the expression `UCP_DT_IS_CONTIG(sreq->send.datatype) && ucp_ep_is_rndv_lane_present(sreq->send.ep)`, which is the previous condition of if buffer was registered

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: pls take a look at `ucp_request_send_buffer_reg` and `ucp_request_send_buffer_dereg` functions. i added assertions. It means that `ucp_request_send_buffer_dereg` is called w/o `ucp_request_send_buffer_reg` before or there is double deregistration

Comment (yosefe) [inline src/ucp/tag/rndv.c]: i see, so you mean the buffer was registered without going through ucp_request_send_buffer_reg ?
looks like this should be fixed..

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Probably, but i did not find another place where the buffer caught be registered.

Comment (alinask) [inline src/ucp/tag/rndv.c]: what do you mean?
ucp_rndv_prepare_zcopy_send_buffer
and
ucp_rndv_rma_request_send_buffer_dereg
are not called one after the other, only one of them is called.

also, how does MM register the buffer if neither posix nor sysv have a reg func? MM shouldn't be chosen for rndv

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: may be rndv enables thru other TL because MM is an alias for many TLs in gtests?

Comment (yosefe) [inline src/ucp/tag/rndv.c]: @alinask looks like ucp_rndv_rma_request_send_buffer_dereg can be called when the send buffer was never registered, in case of AM based rndv?

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

File: src/ucp/tag/rndv.c (hunk 2 above)
Comment (alinask) [inline src/ucp/tag/rndv.c]: @evgeny-leksikov shm is the alias in gtest, not mm :)
@yosefe ucp_rndv_rma_request_send_buffer_dereg can be called from AM flow but in has protections inside so that only what was registered would be deregistered:
```
if (req->send.state.dt.contig.memh != UCT_MEM_HANDLE_NULL) {
     uct_md_mem_dereg(uct_md, state->dt.contig.memh);
}
```

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: @alinask sorry for my short memory :) it was couple days ago.
@yosefe I can remove one extra check from these if it's expected, but it's confusing me a bit..

Comment (alinask) [inline src/ucp/tag/rndv.c]: @evgeny-leksikov did you see a failure becuase of which you added this line? I think we need to figure out how this flow happens since this line is a workaround

Comment (yosefe) [inline src/ucp/tag/rndv.c]: @evgeny-leksikov you are getting the failure because you added assertion, which fails before getting to the check @alinask mentioned

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: @yosefe Exactly. I dislike non paired manipulations with resources )
If it's expected i would keep the check outside of dereg call.

Comment (yosefe) [inline src/ucp/tag/rndv.c]: so in this case, for consistency, need to change all places which set memh=NULL to set rsc=NULL, and replace the memh!=NULL by a check of rsc!=NULL

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3778/ for details (Mellanox internal link).

File: src/ucp/tag/rndv.c (hunk 2 above)
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: It's no mater for me which flag to check, i propose move it outside.
lets remove assertions in this PR then do refactoring.

Comment (yosefe) [inline src/ucp/tag/rndv.c]: you can't remove assertion because uct_mds[rsc_index] is used
no choice but to keep the check as you did already, and add TODO with the explanation of the problem

Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: ok

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1845/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3787/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3927/ for details (Mellanox internal link).

File: src/ucp/core/ucp_request.h [hunk 3]
```diff
@@ -140,6 +140,7 @@ struct ucp_request {
             };
 
             ucp_lane_index_t      lane;     /* Lane on which this request is being sent */
+            ucp_rsc_index_t       rsc;      /* Resource on which memory is registered */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: maybe rename to "reg_rsc"?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3931/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3936/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by MattBBaker** [comment]
Seg fault,
```
[----------] 1 test from shm/test_ucp_peer_failure_zcopy
[ RUN      ] shm/test_ucp_peer_failure_zcopy.status_after_error/0
[pod9:21188:0] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x4000000200)
[pod9:21188:0] Process frozen...
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3947/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
bot:bgate:retest

**Comment by yosefe** [comment]
@shamisp pls review as well

**Comment by shamisp** [comment]
👍
